{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import kipoi\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Source Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "0.00B [00:00, ?B/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading https://zenodo.org/record/4075690/files/humanMedian_trainepoch.11-0.426.h5 to /data/nasif12/home_if12/karollus/5UTRModel/xpresso/Xpresso_kipoi/Xpresso_kipoi/downloaded/model_files/human_median/weights/9d00a3bc614da81655328b6e278569e2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1.39MB [00:00, 1.84MB/s]                            "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py:263: colocate_with (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Colocations handled automatically by placer.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:3445: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.\n",
      "WARNING:tensorflow:From /opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/tensorflow/python/ops/math_ops.py:3066: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "Use tf.cast instead.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/modules/i12g/anaconda/envs/kipoi-Framepool2/lib/python3.6/site-packages/keras/engine/saving.py:327: UserWarning: Error in loading the saved optimizer state. As a result, your model is starting with a freshly initialized optimizer.\n",
      "  warnings.warn('Error in loading the saved optimizer '\n"
     ]
    }
   ],
   "source": [
    "# Source model directly from directory\n",
    "model = kipoi.get_model(\"../Xpresso_kipoi/human_median\", source=\"dir\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download and prepare example files (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "import gzip\n",
    "import shutil\n",
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make ExampleFile directory if it does not exist\n",
    "if not os.path.exists(\"ExampleFiles\"):\n",
    "    os.makedirs(\"ExampleFiles\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('ExampleFiles/chrom22.fa', <http.client.HTTPMessage at 0x2b89ab7aefd0>)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Download GTF\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/1466102/files/example_files-gencode.v24.annotation_chr22.gtf?download=1\", 'ExampleFiles/chrom22.gtf')\n",
    "# Download fasta\n",
    "urllib.request.urlretrieve(\"https://zenodo.org/record/1466102/files/example_files-hg38_chr22.fa?download=1\", 'ExampleFiles/chrom22.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract implied TSS sites from gtf\n",
    "# Read in with pyranges\n",
    "gr = pr.read_gtf('ExampleFiles/chrom22.gtf')\n",
    "# Extract protein coding genes\n",
    "prot_genes = gr.df[(gr.df.Feature == 'gene') & (gr.df.gene_type == 'protein_coding')]\n",
    "# Compute implied TSS\n",
    "prot_genes['TSS'] = (prot_genes.Start * (prot_genes.Strand == \"+\")) + (prot_genes.End * (prot_genes.Strand == \"-\"))\n",
    "# Determine region around TSS\n",
    "prot_genes['region_start'] = prot_genes.TSS + (-7000*(prot_genes.Strand == \"+\")) + (-3500 * (prot_genes.Strand == \"-\"))\n",
    "prot_genes['region_end'] = prot_genes.TSS + (3500*(prot_genes.Strand == \"+\")) + (7000 * (prot_genes.Strand == \"-\"))\n",
    "# Add nuisance column to make bed6\n",
    "prot_genes[\"score\"] = \".\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write bed file\n",
    "bed = prot_genes[['Chromosome', 'region_start', 'region_end', 'gene_id', 'score', 'Strand']]\n",
    "bed.to_csv(\"ExampleFiles/chrom22.bed\", sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Provide the Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path of the fasta file\n",
    "fasta_path = \"ExampleFiles/chrom22.fa\"\n",
    "# Set false if fasta has a chr prefix, true otherwise\n",
    "num_chr = False\n",
    "\n",
    "# Path of the bed file specifying the promoter regions\n",
    "bed_path = \"ExampleFiles/chrom22.bed\"\n",
    "\n",
    "# output file path\n",
    "output_file_path = \"predictions.tsv\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Prediction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7/7 [00:16<00:00,  2.42s/it]\n"
     ]
    }
   ],
   "source": [
    "model.pipeline.predict_to_file(output_file_path, {\"intervals_file\":bed_path, \n",
    "                               \"fasta_file\":fasta_path,\n",
    "                               \"num_chr_fasta\":num_chr},\n",
    "                              batch_size=64)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metadata/ranges/chr</th>\n",
       "      <th>metadata/ranges/end</th>\n",
       "      <th>metadata/ranges/id</th>\n",
       "      <th>metadata/ranges/start</th>\n",
       "      <th>metadata/ranges/strand</th>\n",
       "      <th>preds/expression_pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22</td>\n",
       "      <td>11070000</td>\n",
       "      <td>0</td>\n",
       "      <td>11059500</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.097702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22</td>\n",
       "      <td>15531657</td>\n",
       "      <td>1</td>\n",
       "      <td>15521157</td>\n",
       "      <td>+</td>\n",
       "      <td>-1.052358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr22</td>\n",
       "      <td>15693525</td>\n",
       "      <td>2</td>\n",
       "      <td>15683025</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.806651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22</td>\n",
       "      <td>17088453</td>\n",
       "      <td>3</td>\n",
       "      <td>17077953</td>\n",
       "      <td>+</td>\n",
       "      <td>0.872204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22</td>\n",
       "      <td>17363448</td>\n",
       "      <td>4</td>\n",
       "      <td>17352948</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.903286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50585465</td>\n",
       "      <td>434</td>\n",
       "      <td>50574965</td>\n",
       "      <td>-</td>\n",
       "      <td>0.178786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50589965</td>\n",
       "      <td>435</td>\n",
       "      <td>50579465</td>\n",
       "      <td>-</td>\n",
       "      <td>0.955409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50608455</td>\n",
       "      <td>436</td>\n",
       "      <td>50597955</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.159009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50635173</td>\n",
       "      <td>437</td>\n",
       "      <td>50624673</td>\n",
       "      <td>-</td>\n",
       "      <td>0.754210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50790663</td>\n",
       "      <td>438</td>\n",
       "      <td>50780163</td>\n",
       "      <td>-</td>\n",
       "      <td>0.078148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>439 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    metadata/ranges/chr  metadata/ranges/end  metadata/ranges/id  \\\n",
       "0                 chr22             11070000                   0   \n",
       "1                 chr22             15531657                   1   \n",
       "2                 chr22             15693525                   2   \n",
       "3                 chr22             17088453                   3   \n",
       "4                 chr22             17363448                   4   \n",
       "..                  ...                  ...                 ...   \n",
       "434               chr22             50585465                 434   \n",
       "435               chr22             50589965                 435   \n",
       "436               chr22             50608455                 436   \n",
       "437               chr22             50635173                 437   \n",
       "438               chr22             50790663                 438   \n",
       "\n",
       "     metadata/ranges/start metadata/ranges/strand  preds/expression_pred  \n",
       "0                 11059500                      +              -0.097702  \n",
       "1                 15521157                      +              -1.052358  \n",
       "2                 15683025                      +              -0.806651  \n",
       "3                 17077953                      +               0.872204  \n",
       "4                 17352948                      +              -0.903286  \n",
       "..                     ...                    ...                    ...  \n",
       "434               50574965                      -               0.178786  \n",
       "435               50579465                      -               0.955409  \n",
       "436               50597955                      -              -0.159009  \n",
       "437               50624673                      -               0.754210  \n",
       "438               50780163                      -               0.078148  \n",
       "\n",
       "[439 rows x 6 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load data as dataframe\n",
    "df = pd.read_csv(output_file_path, sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge back with gene_ids\n",
    "df = df.rename(columns={\"metadata/ranges/chr\":\"Chromosome\", \"metadata/ranges/start\":\"region_start\", \"metadata/ranges/end\":\"region_end\", \"metadata/ranges/strand\":\"strand\"})\n",
    "merged = prot_genes.merge(df, on=[\"Chromosome\", \"region_start\", \"region_end\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>TSS</th>\n",
       "      <th>region_start</th>\n",
       "      <th>region_end</th>\n",
       "      <th>score</th>\n",
       "      <th>metadata/ranges/id</th>\n",
       "      <th>strand</th>\n",
       "      <th>preds/expression_pred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>11066500</td>\n",
       "      <td>11068089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000279973.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11066500</td>\n",
       "      <td>11059500</td>\n",
       "      <td>11070000</td>\n",
       "      <td>.</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.097702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>15528157</td>\n",
       "      <td>15529139</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000130538.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15528157</td>\n",
       "      <td>15521157</td>\n",
       "      <td>15531657</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>-1.052358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>15690025</td>\n",
       "      <td>15721631</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000198062.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15690025</td>\n",
       "      <td>15683025</td>\n",
       "      <td>15693525</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.806651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>17084953</td>\n",
       "      <td>17115694</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000177663.13</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17084953</td>\n",
       "      <td>17077953</td>\n",
       "      <td>17088453</td>\n",
       "      <td>.</td>\n",
       "      <td>3</td>\n",
       "      <td>+</td>\n",
       "      <td>0.872204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>17359948</td>\n",
       "      <td>17558149</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000099954.18</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17359948</td>\n",
       "      <td>17352948</td>\n",
       "      <td>17363448</td>\n",
       "      <td>.</td>\n",
       "      <td>4</td>\n",
       "      <td>+</td>\n",
       "      <td>-0.903286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>434</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50568860</td>\n",
       "      <td>50578465</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000205560.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50578465</td>\n",
       "      <td>50574965</td>\n",
       "      <td>50585465</td>\n",
       "      <td>.</td>\n",
       "      <td>434</td>\n",
       "      <td>-</td>\n",
       "      <td>0.178786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50568868</td>\n",
       "      <td>50582965</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000254413.8</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50582965</td>\n",
       "      <td>50579465</td>\n",
       "      <td>50589965</td>\n",
       "      <td>.</td>\n",
       "      <td>435</td>\n",
       "      <td>-</td>\n",
       "      <td>0.955409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50578948</td>\n",
       "      <td>50601455</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000100288.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50601455</td>\n",
       "      <td>50597955</td>\n",
       "      <td>50608455</td>\n",
       "      <td>.</td>\n",
       "      <td>436</td>\n",
       "      <td>-</td>\n",
       "      <td>-0.159009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50622753</td>\n",
       "      <td>50628173</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000100299.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50628173</td>\n",
       "      <td>50624673</td>\n",
       "      <td>50635173</td>\n",
       "      <td>.</td>\n",
       "      <td>437</td>\n",
       "      <td>-</td>\n",
       "      <td>0.754210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>chr22</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>50767500</td>\n",
       "      <td>50783663</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000079974.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50783663</td>\n",
       "      <td>50780163</td>\n",
       "      <td>50790663</td>\n",
       "      <td>.</td>\n",
       "      <td>438</td>\n",
       "      <td>-</td>\n",
       "      <td>0.078148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>439 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome   Source Feature     Start       End Score Strand Frame  \\\n",
       "0        chr22  ENSEMBL    gene  11066500  11068089     .      +     .   \n",
       "1        chr22   HAVANA    gene  15528157  15529139     .      +     .   \n",
       "2        chr22   HAVANA    gene  15690025  15721631     .      +     .   \n",
       "3        chr22   HAVANA    gene  17084953  17115694     .      +     .   \n",
       "4        chr22   HAVANA    gene  17359948  17558149     .      +     .   \n",
       "..         ...      ...     ...       ...       ...   ...    ...   ...   \n",
       "434      chr22   HAVANA    gene  50568860  50578465     .      -     .   \n",
       "435      chr22   HAVANA    gene  50568868  50582965     .      -     .   \n",
       "436      chr22   HAVANA    gene  50578948  50601455     .      -     .   \n",
       "437      chr22   HAVANA    gene  50622753  50628173     .      -     .   \n",
       "438      chr22   HAVANA    gene  50767500  50783663     .      -     .   \n",
       "\n",
       "                gene_id       gene_type  ... havana_transcript protein_id  \\\n",
       "0     ENSG00000279973.1  protein_coding  ...               NaN        NaN   \n",
       "1     ENSG00000130538.4  protein_coding  ...               NaN        NaN   \n",
       "2    ENSG00000198062.14  protein_coding  ...               NaN        NaN   \n",
       "3    ENSG00000177663.13  protein_coding  ...               NaN        NaN   \n",
       "4    ENSG00000099954.18  protein_coding  ...               NaN        NaN   \n",
       "..                  ...             ...  ...               ...        ...   \n",
       "434  ENSG00000205560.12  protein_coding  ...               NaN        NaN   \n",
       "435   ENSG00000254413.8  protein_coding  ...               NaN        NaN   \n",
       "436  ENSG00000100288.19  protein_coding  ...               NaN        NaN   \n",
       "437  ENSG00000100299.17  protein_coding  ...               NaN        NaN   \n",
       "438  ENSG00000079974.17  protein_coding  ...               NaN        NaN   \n",
       "\n",
       "    ccdsid       TSS region_start region_end score metadata/ranges/id strand  \\\n",
       "0      NaN  11066500     11059500   11070000     .                  0      +   \n",
       "1      NaN  15528157     15521157   15531657     .                  1      +   \n",
       "2      NaN  15690025     15683025   15693525     .                  2      +   \n",
       "3      NaN  17084953     17077953   17088453     .                  3      +   \n",
       "4      NaN  17359948     17352948   17363448     .                  4      +   \n",
       "..     ...       ...          ...        ...   ...                ...    ...   \n",
       "434    NaN  50578465     50574965   50585465     .                434      -   \n",
       "435    NaN  50582965     50579465   50589965     .                435      -   \n",
       "436    NaN  50601455     50597955   50608455     .                436      -   \n",
       "437    NaN  50628173     50624673   50635173     .                437      -   \n",
       "438    NaN  50783663     50780163   50790663     .                438      -   \n",
       "\n",
       "    preds/expression_pred  \n",
       "0               -0.097702  \n",
       "1               -1.052358  \n",
       "2               -0.806651  \n",
       "3                0.872204  \n",
       "4               -0.903286  \n",
       "..                    ...  \n",
       "434              0.178786  \n",
       "435              0.955409  \n",
       "436             -0.159009  \n",
       "437              0.754210  \n",
       "438              0.078148  \n",
       "\n",
       "[439 rows x 33 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:anaconda-kipoi-Framepool2]",
   "language": "python",
   "name": "conda-env-anaconda-kipoi-Framepool2-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
